src/time.h

```diff
@@ -146,6 +148,54 @@ class ExpirationImpl : public Expiration {
   typename Clock::time_point token_expiration_;
 };
 
+// Abstract class for a timer that can perform asynchronous actions on
+// expiration.
+class WaitableTimer {
+ public:
+  virtual ~WaitableTimer() = default;
+  virtual void ExpiresFromNow(std::chrono::seconds duration) = 0;
+  virtual void AsyncWait(
+      std::function<void(boost::system::error_code const &)>) = 0;
+};
+
+// Implementation of WaitableTimer parameterized over a clock type.
+template<typename Clock, typename WaitTraits>
+class WaitableTimerImpl : public WaitableTimer {
+ public:
+  WaitableTimerImpl(boost::asio::io_service& service) : timer_(service) {}
+  void ExpiresFromNow(std::chrono::seconds duration) override {
+    timer_.expires_from_now(duration);
+  }
+  void AsyncWait(
+      std::function<void(boost::system::error_code const &)> handler) override {
+    timer_.async_wait(handler);
+  }
+ private:
+  boost::asio::basic_waitable_timer<Clock, WaitTraits> timer_;
+};
+
+// Abstract class for a factory of WaitableTimer types.
+class WaitableTimerFactory {
+ public:
+  virtual ~WaitableTimerFactory() = default;
+  virtual std::unique_ptr<WaitableTimer> CreateTimer(
+      boost::asio::io_service& service) = 0;
+};
+
+// Implementation of a WaitableTimer parameterized over a clock type.
+template<typename Clock, typename WaitTraits = boost::asio::wait_traits<Clock>>
```

Comment:
From Google's C++ style guide, function overloading is preferred to default values for readability.

Recommend either:
1. Add another implementation to template with Clock only.
2. Let the user of the library always specify the wait_traits for their use_case.

I prefer 1. because it lets the user easy to use it correctly and harder to use it incorrectly, which is the intention of current code.